### PR TITLE
[Bug] Fix first-turn status damage and arena hazards

### PR DESCRIPTION
**Pull request by ElizaAlex**
## What are the changes?

- Adjusted MoveEffectPhase.start() so that ENEMY_SIDE targeted moves no longer occur twice per use in double battles.
- Improved documentation / code readability
- Fixed Toxic / Poison dealing damage immediately when applied.
- Fixed Hazards not persisting through save
- Updated Toxic Orb test

I wasn't able to recreate toxic spikes not being absorbed, or the double message, though I didn't attempt to until after fixing the double battle problem. 

## Why am I doing these changes?
Addresses bugs compiled in #3240 

## What did change?
- Adjusted MoveEffectPhase to only apply the effects of ENEMY_SIDE targeted moves a single time, even in double battles. Previously this caused a bug where you would apply two layers of toxic spikes (and presumably spikes)
- Fixed bug where toxin/poison/burn damage would apply prematurely by adding a new phase scheduled after weather effects to check for and queue status damage.
- As a result of the Poison/Toxic/Burn changes, Toxic and Flame Orb now behave as they should, with the damage not triggering until the second turn. 
- Fixed ArenaTags not being loaded properly after a reset by basing them off of the handling of BattlerTags. Previously the ArenaData class constructor never successfully recovered saved tags from the JSON string. 
- Updated Toxic Orb test to reflect changes.
- ~~Refactored MoveEffectPhase in Phases.ts to improve readability, heavily reducing nested code by breaking the section into various documented helper functions~~ Separated out into #3541 

### Screenshots/Videos
Only one layer of Spikes/Toxic Spikes are applied even in double battles:

https://github.com/user-attachments/assets/ad1dc723-475f-453b-9fa8-8de296159438

Only one status affliction message is displayed per target:

https://github.com/user-attachments/assets/5aabe286-e59a-4393-b570-80ebf1fc4ed5

Hazards now persist through a reload:

https://github.com/user-attachments/assets/97e50058-3215-4ad2-b2cb-690f39fb9173

Toxic Spikes are correctly absorbed by grounded poison types:

https://github.com/user-attachments/assets/2c7845ee-9fd4-4928-8c28-20a11e1d439d

Additionally, Pokemon no longer take damage immediately upon being poisoned by Toxic Spikes. (The damage they take upon entering the field is due to the normal Spikes)

## How to test the changes?

1. Use overrides to force double battles, and to give your starter(s) of choice toxic spikes
2. Use toxic spikes in a wild battle, and then defeat your opponent.
3. Proceed far enough into the next encounter so that the incoming pokemon get poisoned. The text should only display once, and only one set should go down.
4. Refresh the game
5. Load back into the same file, and the spikes should still be present / affect the incoming pokemon
6. Repeat steps 2-5 to test the Badly Poisoned effect
7. If at any point a grounded poison type pokemon enters the field, the spikes should be absorbed.

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/phases.ts [hunk 1]
```diff
@@ -3744,9 +3777,6 @@ export class ObtainStatusEffectPhase extends PokemonPhase {
         pokemon.updateInfo(true);
         new CommonBattleAnim(CommonAnim.POISON + (this.statusEffect! - 1), pokemon).play(this.scene, () => {
           this.scene.queueMessage(getStatusEffectObtainText(this.statusEffect, getPokemonNameWithAffix(pokemon), this.sourceText ?? undefined));
-          if (pokemon.status?.isPostTurn()) {
```
Comment (ElizaAlex) [inline src/phases.ts:3747]: This caused premature application of poison/burn damage, such as taking damage turn 1 from Toxic/Poison orb, and taking damage immediately upon landing in toxic spikes.

**Comment by ElizaAlex** [comment]
Updated to reflect removal of Phases.ts

**Comment by Snailman11** [comment]
Probably fixes https://github.com/pagefaultgames/pokerogue/issues/684

File: src/phases/check-status-effect-phase.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import {PostTurnStatusEffectPhase} from "#app/phases/post-turn-status-effect-phase";
+import {Phase} from "#app/phase";
+import {BattlerIndex} from "#app/battle";
```
Comment (DayKev) [inline src/phases/check-status-effect-phase.ts:3]:
```suggestion
import { PostTurnStatusEffectPhase } from "#app/phases/post-turn-status-effect-phase";
import { Phase } from "#app/phase";
import { BattlerIndex } from "#app/battle";
```

File: src/phases/move-effect-phase.ts [hunk 3]
```diff
@@ -14,6 +14,7 @@ import { PokemonMultiHitModifier, FlinchChanceModifier, EnemyAttackStatusEffectC
 import i18next from "i18next";
 import * as Utils from "#app/utils.js";
 import { PokemonPhase } from "./pokemon-phase";
+import {Abilities} from "#enums/abilities";
```
Comment (DayKev) [inline src/phases/move-effect-phase.ts:17]:
```suggestion
import { Abilities } from "#enums/abilities";
```

File: src/phases/turn-start-phase.ts [hunk 4]
```diff
@@ -13,10 +13,10 @@ import { BerryPhase } from "./berry-phase";
 import { FieldPhase } from "./field-phase";
 import { MoveHeaderPhase } from "./move-header-phase";
 import { MovePhase } from "./move-phase";
-import { PostTurnStatusEffectPhase } from "./post-turn-status-effect-phase";
 import { SwitchSummonPhase } from "./switch-summon-phase";
 import { TurnEndPhase } from "./turn-end-phase";
 import { WeatherEffectPhase } from "./weather-effect-phase";
+import {CheckStatusEffectPhase} from "#app/phases/check-status-effect-phase";
```
Comment (DayKev) [inline src/phases/turn-start-phase.ts:19]:
```suggestion
import { CheckStatusEffectPhase } from "#app/phases/check-status-effect-phase";
```

File: src/system/arena-data.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import { Arena } from "../field/arena";
-import { ArenaTag } from "../data/arena-tag";
+import {ArenaTag, loadArenaTag} from "../data/arena-tag";
```
Comment (DayKev) [inline src/system/arena-data.ts:2]:
```suggestion
import { ArenaTag, loadArenaTag } from "../data/arena-tag";
```

File: src/system/game-data.ts [hunk 6]
```diff
@@ -44,6 +44,8 @@ import { WeatherType } from "#app/enums/weather-type.js";
 import { TerrainType } from "#app/data/terrain.js";
 import { OutdatedPhase } from "#app/phases/outdated-phase.js";
 import { ReloadSessionPhase } from "#app/phases/reload-session-phase.js";
+import {TagAddedEvent} from "#app/events/arena";
+import {ArenaTrapTag} from "#app/data/arena-tag";
```
Comment (DayKev) [inline src/system/game-data.ts:48]:
```suggestion
import { TagAddedEvent } from "#app/events/arena";
import { ArenaTrapTag } from "#app/data/arena-tag";
```

**Comment by DayKev** [comment]
Merged in beta branch and updated the tests (plus applied some style nits).

File: src/phases/move-effect-phase.ts [hunk 7]
```diff
@@ -123,6 +124,11 @@ export class MoveEffectPhase extends PokemonPhase {
       new MoveAnim(move.id as Moves, user, this.getTarget()?.getBattlerIndex()!).play(this.scene, () => { // TODO: is the bang correct here?
         /** Has the move successfully hit a target (for damage) yet? */
         let hasHit: boolean = false;
+
+        /** Hazards bounce if any opposing Pokémon has the ability **/
+        /** TODO: Update to support future Magic Coat / Magic Bounce implementation **/
+        let hazardBounce: boolean = false;
```
Comment (innerthunder) [inline src/phases/move-effect-phase.ts:130]: The Magic Bounce-related stuff is a nice thought, but I don't really think it fits within the scope of this PR

**Review by innerthunder — CHANGES_REQUESTED** [review]

File: src/phases/move-effect-phase.ts [hunk 8]
```diff
@@ -123,6 +123,7 @@ export class MoveEffectPhase extends PokemonPhase {
       new MoveAnim(move.id as Moves, user, this.getTarget()?.getBattlerIndex()!).play(this.scene, () => { // TODO: is the bang correct here?
         /** Has the move successfully hit a target (for damage) yet? */
         let hasHit: boolean = false;
+
```
Comment (DayKev) [inline src/phases/move-effect-phase.ts:126]:
```suggestion
```

File: src/test/moves/safeguard.test.ts [hunk 9]
```diff
@@ -120,6 +120,7 @@ describe("Moves - Safeguard", () => {
     const enemyPokemon = game.scene.getEnemyPokemon()!;
 
     game.move.select(Moves.SPLASH);
+    enemyPokemon.damageAndUpdate(1);
```
Comment (DayKev) [inline src/test/moves/safeguard.test.ts:123]:
```suggestion
```

**Review by DayKev — APPROVED** [review]

**Review by innerthunder — DISMISSED** [review]

**Review by DayKev — APPROVED** [review]

**Review by DayKev — APPROVED** [review]

**Review by innerthunder — APPROVED** [review]

**Pull request merged**